### ISSUE TITLE
Error in Style/RedundantParentheses

### ISSUE THREAD

**Issue by lordofthelake** [bbatsov/rubocop#3154]
## Expected behavior
Rubocop should just lint the file and not throw an error.

## Actual behavior
An error occurs in the `Style/RedudantParenthesis` cop while linting the file, with the following error message:

```
Inspecting 1 file
An error occurred while Style/RedundantParentheses cop was inspecting /path/to/test.rb.
To see the complete backtrace run rubocop -d.
.

1 file inspected, no offenses detected

1 error occurred:
An error occurred while Style/RedundantParentheses cop was inspecting /path/to/test.rb.
Errors are usually caused by RuboCop bugs.
Please, report your problems to RuboCop's issue tracker.
Mention the following information in the issue report:
0.40.0 (using Parser 2.3.1.0, running on ruby 2.3.1 x86_64-darwin15)
```

## Steps to reproduce the problem

1. Create a test file with the following content:
    ```ruby
    def foo(bar)
      (bar || ()).nil?
    end
    ```  

2. Run `rubocop test.rb`

## RuboCop version

```
$ rubocop -V
0.40.0 (using Parser 2.3.1.0, running on ruby 2.3.1 x86_64-darwin15)
```

**Comment by amencarini** [comment]
Just encountered the same issue.
The `-d` run adds the following:

```
Scanning /Users/alessandromencarini/Projects/cc/platform/internal/payments_engine/spec/app/processors/payments/payment_creator_spec.rb
An error occurred while Style/RedundantParentheses cop was inspecting /Users/alessandromencarini/Projects/cc/platform/internal/payments_engine/spec/app/processors/payments/payment_creator_spec.rb.
undefined method `hash_type?' for nil:NilClass
/Users/alessandromencarini/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubocop-0.40.0/lib/rubocop/cop/style/redundant_parentheses.rb:41:in `parens_allowed?'
/Users/alessandromencarini/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubocop-0.40.0/lib/rubocop/cop/style/redundant_parentheses.rb:30:in `on_begin'
/Users/alessandromencarini/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubocop-0.40.0/lib/rubocop/cop/commissioner.rb:42:in `block (2 levels) in on_begin'
```
